File: backend/app/schemas/progress.py
from datetime import datetime

from pydantic import BaseModel


# ============================================================
# Lesson Progress
# ============================================================

class LessonProgressResponse(BaseModel):

    lesson_id: int

    lesson_title: str

    status: str

    progress_percentage: int

    completed_at: datetime | None


# ============================================================
# Quiz Progress
# ============================================================

class QuizProgressResponse(BaseModel):

    quiz_id: int

    quiz_title: str

    attempts: int

    best_score: int

    total_points: int

    best_percentage: float

    passed: bool


# ============================================================
# Simulation Progress
# ============================================================

class SimulationProgressResponse(BaseModel):

    simulation_id: int

    simulation_title: str

    scenarios_answered: int

    score: int

    max_score: int

    percentage: float

    completed: bool


# ============================================================
# Student Dashboard
# ============================================================

class StudentDashboardResponse(BaseModel):

    total_lessons: int

    completed_lessons: int

    lesson_completion_percentage: float

    total_quizzes: int

    completed_quizzes: int

    quiz_average_percentage: float

    quizzes_passed: int

    total_simulations: int

    completed_simulations: int

    simulation_average_percentage: float

    overall_preparedness_percentage: float

File: frontend/src/api/progress.ts
import apiClient from "./client";

export interface StudentDashboardResponse {
    total_lessons: number;
    completed_lessons: number;
    lesson_completion_percentage: number;

    total_quizzes: number;
    completed_quizzes: number;
    quiz_average_percentage: number;
    quizzes_passed: number;

    total_simulations: number;
    completed_simulations: number;
    simulation_average_percentage: number;

    overall_preparedness_percentage: number;
}

export interface QuizProgressResponse {
    quiz_id: number;
    quiz_title: string;
    attempts: number;
    best_score: number;
    total_points: number;
    best_percentage: number;
    passed: boolean;
}

export interface SimulationProgressResponse {
    simulation_id: number;
    simulation_title: string;
    scenarios_answered: number;
    score: number;
    max_score: number;
    percentage: number;
    completed: boolean;
}

export interface LessonProgressResponse {
    lesson_id: number;
    lesson_title: string;
    status: string;
    progress_percentage: number;
    completed_at: string | null;
}

export const getStudentDashboard =
    async (): Promise<StudentDashboardResponse> => {
        const response =
            await apiClient.get<StudentDashboardResponse>(
                "/progress/dashboard"
            );

        return response.data;
    };

export const getLessonProgress =
    async (): Promise<LessonProgressResponse[]> => {
        const response =
            await apiClient.get<LessonProgressResponse[]>(
                "/progress/lessons"
            );

        return response.data;
    };

export const getQuizProgress =
    async (): Promise<QuizProgressResponse[]> => {
        const response =
            await apiClient.get<QuizProgressResponse[]>(
                "/progress/quizzes"
            );

        return response.data;
    };

export const getSimulationProgress =
    async (): Promise<SimulationProgressResponse[]> => {
        const response =
            await apiClient.get<SimulationProgressResponse[]>(
                "/progress/simulations"
            );

        return response.data;
    };